exchanges/bittrex/bittrex_websocket.go

```diff
@@ -0,0 +1,603 @@
+		Result:        result,
+		Verbose:       b.Verbose,
+		HTTPDebugging: b.HTTPDebugging,
+		HTTPRecording: b.HTTPRecording,
+	})
+}
+
+// WsAuth sends an authentication message to receive auth data
+// Authentications expire after 10 minutes
+func (b *Bittrex) WsAuth() error {
+	// [apiKey, timestamp in ms, random uuid, signed payload]
+	apiKey := b.API.Credentials.Key
+	randomContent, err := uuid.NewV4()
+	if err != nil {
+		return err
+	}
+	timestamp := strconv.FormatInt(time.Now().UnixNano()/1000000, 10)
+	hmac := crypto.GetHMAC(
+		crypto.HashSHA512,
+		[]byte(timestamp+randomContent.String()),
+		[]byte(b.API.Credentials.Secret),
+	)
+	signature := crypto.HexEncodeToString(hmac)
+
+	invocationIDCounter++
+	req := WsEventRequest{
+		Hub:          "c3",
+		Method:       authenticate,
+		InvocationID: invocationIDCounter,
+	}
+
+	arguments := make([]string, 0)
+	arguments = append(arguments, apiKey, timestamp, randomContent.String(), signature)
+	req.Arguments = arguments
+
+	requestString, err := json.Marshal(req)
+	if err != nil {
+		return err
+	}
+	if b.Verbose {
+		log.Debugf(log.WebsocketMgr, "%s Sending JSON message - %s\n", b.Name, requestString)
+	}
+	err = b.Websocket.Conn.SendJSONMessage(req)
+	if err != nil {
+		return err
+	}
+
+	b.WsPendingRequests[req.InvocationID] = WsPendingRequest{
+		WsEventRequest: req,
+	}
+	return nil
+}
+
+// GenerateDefaultSubscriptions Adds default subscriptions to websocket to be
+// handled by ManageSubscriptions()
+func (b *Bittrex) GenerateDefaultSubscriptions() ([]stream.ChannelSubscription, error) {
+	var subscriptions []stream.ChannelSubscription
+	pairs, err := b.GetEnabledPairs(asset.Spot)
+	if err != nil {
+		return nil, err
+	}
+
+	channels := defaultSpotSubscribedChannels
+	if b.GetAuthenticatedAPISupport(exchange.WebsocketAuthentication) {
+		channels = append(channels, defaultSpotSubscribedChannelsAuth...)
+	}
+
+	for i := range pairs {
+		pair, err := b.FormatExchangeCurrency(pairs[i], asset.Spot)
+		if err != nil {
+			return nil, err
+		}
+		for y := range channels {
+			var channel string
+			switch channels[y] {
+			case wsOrderbook:
+				channel = channels[y] + "_" + pair.String() + "_" + strconv.FormatInt(orderbookDepth, 10)
+			case wsTicker:
+				channel = channels[y] + "_" + pair.String()
+			case wsMarketSummary:
+				channel = channels[y] + "_" + pair.String()
+			default:
+				channel = channels[y]
+			}
+			subscriptions = append(subscriptions,
+				stream.ChannelSubscription{
+					Channel:  channel,
+					Currency: pair,
+					Asset:    asset.Spot,
+				})
+		}
+	}
+
+	return subscriptions, nil
+}
+
+// Subscribe sends a websocket message to receive data from the channel
+func (b *Bittrex) Subscribe(channelsToSubscribe []stream.ChannelSubscription) error {
+	invocationIDCounter++
+	req := WsEventRequest{
+		Hub:          "c3",
+		Method:       subscribe,
+		InvocationID: invocationIDCounter,
+	}
+
+	var channels []string
+	for i := range channelsToSubscribe {
```

Comment:
This function has the same issues as being solved in another PR: #670 
If one has too many currencies enabled (in this case > 500 in a request), bittrex will forcefully close the connection. This can be resolved in the same way the linked PR does. This would likely also apply to the unsubscribe function.

It may also have a need for a ratelimit as well. See this comment for reference: https://github.com/thrasher-corp/gocryptotrader/pull/670/files#r622624692

To verify, you can run `go run . -enableallpairs`

Reply:
Hm ok I now added a commit to do something similar, but enabling that many currencies stays problematic..
- Not all enabled pairs appear to be available as orderbook streams (using `getorderbookstream`)
- I'm getting this: `[WARN]  | 03/05/2021 10:17:53 | Bittrex websocket: has not received a traffic alert in 30s. Reconnecting`
- Since most books are low liquidity, it'll switch to REST after 1 min for sure

Keep it this way, revert, or something different?

Reply:
![Screenshot-orderbook](https://user-images.githubusercontent.com/68383301/116860983-ae5a5600-ac02-11eb-8672-5593b0df534c.png)
Could it be that there's a limit to the number of orderbooks that are synchronized? 20? If so, then I guess it's OK that I'm not seeing all streams..

Reply:
Hm never mind, it's fuller now, perhaps it just took a while to populate..